test: pytest-harness: post-script run sequence is wrong in pytest harness

### Describe the bug

after below commit
commit 1b7810f0cb74bae33f074573c1b7b6f3d3d2cbf3 (HEAD -> b1b7810f0cb74bae33f074573c1b7b6f3d3d2cbf3, refs/bisect/bad)
Author: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>
Date:   Mon Nov 24 17:13:25 2025 +0100

    scripts: twister: Support multiple ports in pytest-twister-harness

    Add support for multiple connections (e.g. UART) from one device,
    enabling communication with second core UARTs. Feature implemented
    by extracting connection logic from DeviceAdapter to new
    DeviceConnection class with specialized implementations:
    - SerialConnection for hardware UART ports
    - ProcessConnection for native simulation
    - FifoConnection for QEMU communication
    Each connection maintains separate log files (handler.log,
    handler_1.log, etc.) and can be accessed via connection_index
    parameter in device methods like readline() and write().
    This enables testing multi-core applications where different
    cores communicate through separate UART interfaces.

    Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

the post-scripts run sequence is wrong.

### Regression

- [x] This is a regression.

### Steps to reproduce

add any post_script in map file.

Modify the as below
--- a/scripts/pylib/pytest-twister-harness/src/twister_harness/device/hardware_adapter.py
+++ b/scripts/pylib/pytest-twister-harness/src/twister_harness/device/hardware_adapter.py
@@ -178,6 +178,7 @@ class HardwareAdapter(DeviceAdapter):

     def _close_device(self) -> None:
         if self.device_config.post_script:
+            asdfsfsdf
             self._run_custom_script(self.device_config.post_script, self.base_timeout)

and run the 

scripts/twister --device-testing --hardware-map ${HOME}/nxp/frdm_rw612/map.yaml -T samples/drivers/flash_shell/


### Relevant log output

```shell
the failure happens as below sequence, before we see the pytest run ends
pytest --twister-harness -s -v --log-level=DEBUG --build-dir=/home/ubuntu/zephyrproject/zephyr/twister-out/frdm_rw612_rw612/zephyr/samples/drivers/flash_shell/sample.drivers.flash.shell --junit-xml=/home/ubuntu/zephyrproject/zephyr/twister-out/frdm_rw612_rw612/zephyr/samples/drivers/flash_shell/sample.drivers.flash.shell/report.xml --platform=frdm_rw612/rw612 /home/ubuntu/zephyrproject/zephyr/scripts/pylib/shell-twister-harness --base-timeout=60 --device-type=hardware --device-serial=/dev/serial/by-id/usb-NXP_Semiconductors_MCU-LINK_on-board__r0E7__CMSIS-DAP_V3.148_KDWVVIUO1REQK-if02 --device-serial-baud=115200 --flash-timeout=60 --runner=linkserver --device-id=KDWVVIUO1REQK --device-product=LinkServer --post-script=/home/ubuntu/nxp/frdm_rw612/program_hello_world.sh --twister-fixture=spi_loopback --twister-fixture=gpio_loopback --twister-fixture=hsgpio_loopback --testdata=/home/ubuntu/zephyrproject/zephyr/samples/drivers/flash_shell/test_shell.yml -p twister_harness.plugin

============================= test session starts ==============================
platform linux -- Python 3.10.12, pytest-8.1.1, pluggy-1.4.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/ubuntu/zephyrproject/zephyr
collecting ... collected 1 item
scripts/pylib/shell-twister-harness/test_shell.py::test_shell_harness ERROR
scripts/pylib/shell-twister-harness/test_shell.py::test_shell_harness ERROR
==================================== ERRORS ====================================
_____________________ ERROR at setup of test_shell_harness _____________________
request = <SubRequest 'dut' for <Function test_shell_harness>>
device_object = <twister_harness.device.hardware_adapter.HardwareAdapter object at 0x748613ea5030>
    @pytest.fixture(scope=determine_scope)
    def dut(request: pytest.FixtureRequest, device_object: DeviceAdapter) -> Generator[DeviceAdapter, None, None]:
        """Return launched device - with run application."""
        device_object.initialize_log_files(request.node.name)
        try:
>           device_object.launch()
scripts/pylib/pytest-twister-harness/src/twister_harness/fixtures.py:66:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
scripts/pylib/pytest-twister-harness/src/twister_harness/device/device_adapter.py:63: in launch
    self.close()
scripts/pylib/pytest-twister-harness/src/twister_harness/device/device_adapter.py:76: in close
    self._close_device()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
self = <twister_harness.device.hardware_adapter.HardwareAdapter object at 0x748613ea5030>
    def _close_device(self) -> None:
        if self.device_config.post_script:
>           asdfsfsdf
E           NameError: name 'asdfsfsdf' is not defined
```

### Impact

Annoyance – Minor irritation; no significant impact on usability or functionality.

### Environment

- OS: ubuntu
- toolchain: Zephyr
- version: v4.3.0-2785-g1b7810f0cb7

### Additional Context

_No response_

--- a/scripts/pylib/pytest-twister-harness/src/twister_harness/device/device_adapter.py
+++ b/scripts/pylib/pytest-twister-harness/src/twister_harness/device/device_adapter.py
@@ -223,3 +223,7 @@ class DeviceAdapter(abc.ABC):
         Added to keep backward compatibility as it is used in fixtures.
         """
         return self.connections and self.connections[0].is_device_connected()
+
+    def is_reader_started(self) -> bool:
+        """Check if the reader thread is started."""
+        return self._reader_started.is_set()

--- a/scripts/pylib/pytest-twister-harness/src/twister_harness/device/hardware_adapter.py
+++ b/scripts/pylib/pytest-twister-harness/src/twister_harness/device/hardware_adapter.py
@@ -177,7 +177,9 @@ class HardwareAdapter(DeviceAdapter):
                 raise TwisterHarnessException(msg)
 
     def _close_device(self) -> None:
-        if self.device_config.post_script:
+        # Run post script only if the reader thread is started to avoid running it
+        # multiple times and when in initialization phase
+        if self.is_reader_started() and self.device_config.post_script:
             self._run_custom_script(self.device_config.post_script, self.base_timeout)
 
     @staticmethod

